### PR TITLE
 usb: sam0: add a USB device driver.

### PR DESCRIPTION
**Pull request by nzmichaelh**
This patch adds a USB device driver for the SAM0 series and enables it on the Arduino Zero.  I'll also enable it for the Trinket M0 once #6210 lands.

Tested against the cdc_acm and netusb examples.

### CODE REVIEW

File: boards/arm/arduino_zero/doc/arduino_zero.rst [hunk 1]
```diff
@@ -82,6 +84,14 @@ SPI Port
 The SAMD21 MCU has 6 SERCOM based SPIs.  On the Arduino Zero, SPI4 is
 available on the 6 pin connector at the edge of the board.
 
+USB Device Port
+===============
+
+The SAMD21 MCU has a USB device port that can be used to communicate
+with a host PC.  See the :ref:`subsys/usb` sample applications for
```
Comment (dbkinder) [inline boards/arm/arduino_zero/doc/arduino_zero.rst]: The :ref: tag takes the label name of the target location you want to link to, for example, the label that's at the top of the sample app README.rst files, so update these :ref: uses to be:

    See the :ref:`usb-samples` sample applications ...
    ... such as the :ref:`usb_cdc-acm` sample, which sets up ...

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: boards/arm/arduino_zero/doc/arduino_zero.rst (hunk 1 above)
Comment (nzmichaelh) [reply boards/arm/arduino_zero/doc/arduino_zero.rst]: Done, thanks for the pointers.

**Comment by ydamigos** [comment]
@nzmichaelh PR #4376 introduces generic usb bindings, could you review it and check if they work with your driver?

**Comment by dbkinder** [comment]
Looks like you're missing a Signed-off-by: line in the commit message.  As explained in the project [contribution guidelines](http://docs.zephyrproject.org/contribute/contribute_guidelines.html#contribution-tools-and-git-setup), add your name and email to your git configuration with:

    git config --global user.name "David Developer"
    git config --global user.email "david.developer@company.com" 

And update the commit message to include your signature with:

    git commit --amend -s

Exit the commit message editor (shouldn't need any further changes), and force push the change:

    git commit -f origin upstream_usb

and that should trigger the CI system to retest your PR (and hopefully pass this time).

**Comment by nzmichaelh** [comment]
Thanks David, fixed.

**Review by dbkinder — APPROVED** [review]
Thanks!

**Comment by nzmichaelh** [comment]
@ydamigos I commented on the PR.  The concept LGTM.

**Comment by henrikbrixandersen** [comment]
I have successfully tested this on an Adafruit Feather M0 Basic Proto (see #6310) using the `cdc_acm` sample. Thanks!

**Comment by nzmichaelh** [comment]
@ydamigos I've rebased on master and updated the Device Tree binding to use usb-ep.yaml.

**Comment by nzmichaelh** [comment]
I've also tested this with the ECM driver using flood ping and it works fine, yay!

File: dts/bindings/usb/atmel,sam0-usb.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    Atmel SAM0 USB in device mode
+
+inherits:
+    !include usb-ep.yaml
+
+properties:
+    compatible:
+      type: string
+      category: required
+      description: compatible strings
```
Comment (ydamigos) [inline dts/bindings/usb/atmel,sam0-usb.yaml]: we don't need type, category, description. They are already declared in usb.yaml included by usb-ep.yaml

File: drivers/usb/device/usb_dc_sam0.c [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+#define NVM_USB_PAD_TRANSP_SIZE 5
+#define NVM_USB_PAD_TRIM_POS 55
+#define NVM_USB_PAD_TRIM_SIZE 3
+
+#define USB_SAM0_IN_EP 0x80
+
+#define REGS ((Usb *)CONFIG_USB_DC_SAM0_BASE_ADDRESS)
+
+struct usb_sam0_data {
+	UsbDeviceDescriptor descriptors[USB_EPT_NUM];
```
Comment (ydamigos) [inline drivers/usb/device/usb_dc_sam0.c]: You could also get endpoint number from DT instead of ASF.

**Review by ydamigos — COMMENTED** [review]
Thanks for the changes. Some comments regarding DT usage.

File: dts/bindings/usb/atmel,sam0-usb.yaml (hunk 2 above)
Comment (nzmichaelh) [reply dts/bindings/usb/atmel,sam0-usb.yaml]: Done

File: drivers/usb/device/usb_dc_sam0.c (hunk 3 above)
Comment (nzmichaelh) [reply drivers/usb/device/usb_dc_sam0.c]: Done

**Comment by nzmichaelh** [comment]
I've also turned this on for the Trinket M0 now that the board support has landed.

**Review by ydamigos — APPROVED** [review]
Thanks for the changes. It looks fine regarding DT usage. I don't have the hw to test the driver.

**Comment by fallrisk** [comment]
You are missing an implementation of the function `usb_dc_ep_halt`.

**Comment by nzmichaelh** [comment]
Hi @fallrisk I omitted it as the hardware doesn't have the concept of 'halt', just stall.  It's also not used anywhere in Zephyr and isn't implemented in the stm32 driver.

I'd rather delete it globally than implement it.  WDYT?

**Comment by fallrisk** [comment]
When an endpoint is in HALT state it always returns STALL. The endpoint will do this until it receives a control request to move the endpoint back to IDLE. So I think that if you can set a STALL you can support HALT. However, I don't see how the core driver (usb_device.c) returns the endpoint back to IDLE, so the endpoint would be stuck in HALT. 

Part of the problem is that it looks to me like a HALT state is indefinite responses of STALL on that endpoint until the HALT state is cleared (endpoint returned to IDLE). However, most of the USB peripherals I've seen support just sending one STALL and once it's sent the STALL is cleared leaving it to the firmware programmer to handle sending a STALL every packet after that on the endpoint if in a HALT state. However, the host is supposed to cancel all pending transmissions to the HALTed endpoint if the host received a STALL (a.k.a. functional STALL, Jan Axelson).

Look at the core driver (usb_device.c) and bring your attention to the function usb_handle_std_endpoint_req. When that function receives a control request to clear or set the endpoint halt feature it just calls the corresponding dc stall functions. I currently don't see the purpose of usb_dc_ep_halt if the driver doesn't really support it.  

I think it should return "-ENOSYS", which stands for "Function not implemented," in z/lib/libc/minimal/include/errno.h.

Most of this is coming from reading Jan Axelson's book "USB Complete: The Developer's Guide 4ed". I am currently trying to improve my understanding of HALT and STALL so that I can finish my USB driver. So please feel free to correct me, and forgive the untidy notes.

**Comment by nzmichaelh** [comment]
Thanks for the well-reasoned comment @fallrisk - I was going to ask what the difference is between repeated stalls and halt.  FYI, for the SAM0 you can put an endpoint into stall mode where it will stall indefinitely.

A nit on the return code: you can only use ENOSYS for system calls which is enforced by the checkstyle script.

Returning EINVAL SGTM, although that does turn a link time error into a runtime error.

**Comment by carlescufi** [comment]
@galak should we consider this for 1.11?

**Comment by nzmichaelh** [comment]
@carlescufi SGTM.  I've been using it with the ACM, ECM, and HID drivers and it's been working well.

**Review by nashif — APPROVED** [review]

**Pull request merged**